pkg/render/manager_route_config.go

```diff
@@ -0,0 +1,416 @@
+// Copyright (c) 2024 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package render
```

Comment:
Wonder if now is a good time to make a sub-package for this. `pkg/render` is pretty big at this point.

```
pkg/render/manager/routebuilder
-> route_config_builder.go
-> route_config.go
-> utils.go
```

Or something like that?

Reply:
Can leave `pkg/render/manager.go` where it is for now to avoid expanding scope, but this new code feels like a perfect fit for its own tightly scoped package.

Reply:
I was thinking of sub packaging this, but in general the sub packages reference the parent package and we need to do the reverse. The common package might be an exception, but bi directional importing between sub packages and parent packages can get kind of confusing.

This said, ideally we'd move everything into it's own subpackaging so if it is confusing then it just puts more pressure on putting in some extra work to subpackage the other items.

Reply:
What would be the thing causing the import loop in this case?

Reply:
It wouldn't be causing an import loop, just possible confusion. I find package structures where the parent imports some of it's subpackages and other subpackages import the parent lead to confusing patterns and add the potential for import loops. What if you need a subpacakge that follows the patterns of two of it's sibling packages, one that imports from the parent and one that's imported by the parent.

Best to try to adhere to a single import directly IMHO.

Reply:
Mmmm, I'm not quite following that, and it might just be because I'm having trouble identifying what the various sub packages might be.... butt his feels like a very tightly defined library with an interface and private implementation code, and that probably only needs to be used in one particular place. If the concern is "there might be import loops in the future" then I think those can just be sorted out as they arise.

Seems very reasonable to have this in a sub-package.

Reply:
So I moved it to a sub package and we can just move the other manager stuff there later if there are issues.

> If the concern is "there might be import loops in the future" then I think those can just be sorted out as they arise.

I still wanted to comment on this though, these issue can be avoided by having a single import direction and standard with your sub packages, i.e. only allowing imports from one direction parent to child or child to parent or having clear exceptions for certain certain packages. If you run into issues because of this, sorting it out becomes the responsibility of the person who added the last bit of code that broke it and could put re organising multiple packages to adhere to a single direction. I've been somebody whose had to deal with these issues before, and I had thought it was general good golang practice to define such standards for you packages to avoid this.

Reply:
I'm not aware of any such standard / recommendation (it might exist, I'm just not aware of it!)

That "only import below" seems overly limiting in my opinion - package import loops are rare in my experience and usually are resolved by creating a new package as needed. I still struggle to see how this particular package has a high chance of causing problems in the future. It's only going to be imported by `pkg/render`, which is one package above it anyway?

Reply:
> That "only import below" seems overly limiting in my opinion

I wasn't saying it's "only import below", although I tend to have my subpackages import parent packages (since normally the parents contain common interfacing that the children packages take advantage of or are implementers of), it depends on the purposes of the subpackages.

In this instance, most of these packages (with the exception of the common package) import the render parent package for utilities. The manager package would house such similar logic (which is where I moved this configuration too). So if we add more utilities for the manager that actually renders manager logic then we can't add it to this package unless we move the rest of the manager logic there (since it probably needs to import from the render package). We could separate out this specific configuration into it's specific voltron route configuration package, but it doesn't seem like it really needs to be completely isolated (we don't need to create a subpackage for every type).

So sure there are some simple solution to package import loops, but I'd say it's probably a good architectural practice to understand the general relationship between the parent packages and lay down some principles to start with. I have run into issues where I've put this in the wrong direction, mainly for database interface implementations and I wasn't able to switch the direction without some refactoring (debatable whether or not it was trivial), so I take this into consideration any time I have to deal with parent / subpackage interactions.